### PR TITLE
[Crypto] performance improvement of SSWU map to curve

### PR DESCRIPTION
**Pull request by tarakby**
The crypto library has been using Relic's map-to-curve implementation of [SSWU algorithm](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-6.6.2).

Specific optimizations could be implemented in the case of mapping to BLS12 381 (A == 0, and p mod 4 = 3) to improve performance. The optimizations are detailed in the IETF draft ([optimized SSWU](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-14#appendix-F.2) and [square root ratio](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-14#appendix-F.2.1.2)).

This PR implements the optimized SSWU for BLS12 381 by updating an older implementation of SSWU (map [coefficient z](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-14#appendix-H.2) updated from -1 to 11 as per the IETF draft) as well as optimizing the computation steps. 

Other updates:
- remove hardcoded constants and rely on Relic for SSWU and BLS12 381 constants (z, isogeny map, prime field).
- cache extra constants required for SSWU (isogeny map constants, Montgomery constants)
- separate bench function for scalar multiplication of generator and generic points
- add a new test vector from the IETF draft 

Bench results on my laptop show about 35% improvement:
```
name       old time/op  new time/op  delta
MapToG1-8   364µs ± 1%   232µs ± 1%  -36.14%  (p=0.000 n=30+29)
```

### CODE REVIEW

**Review by AlexHentschel — APPROVED** [review]

File: crypto/bls12381_hashtocurve.c [hunk 1]
```diff
@@ -142,95 +72,108 @@ const uint64_t iso_Ny_data[ELLP_Ny_LEN][Fp_DIGITS] = {
...
-    fp_sqr(fp_tmp[5], fp_tmp[2]);              // N^2
-    fp_mul(fp_tmp[5], fp_tmp[5], fp_tmp[2]);  // N^3
-    fp_add(fp_tmp[4], fp_tmp[4], fp_tmp[5]);   // U
+    fp_sqr(fp_tmp[2], fp_tmp[3]);                        // N^2
+    fp_sqr(fp_tmp[0], p->z);                             // D^2
+    fp_mul(fp_tmp[4], *a1, fp_tmp[0]);          // a * D^2
+    fp_add(fp_tmp[2], fp_tmp[4], fp_tmp[2]);             // N^2 + a * D^2
+    fp_mul(fp_tmp[2], fp_tmp[3], fp_tmp[2]);             // N^3 + a * N * D^2
+    fp_mul(fp_tmp[0], fp_tmp[0], p->z);                  // V  =  D^3
+    fp_mul(fp_tmp[4], *b1, fp_tmp[0]);          // b * D^3
```
Comment (SaveTheRbtz) [inline crypto/bls12381_hashtocurve.c:151]: nit: // b * V

File: crypto/bls12381_hashtocurve.c [hunk 2]
```diff
@@ -142,95 +72,108 @@ const uint64_t iso_Ny_data[ELLP_Ny_LEN][Fp_DIGITS] = {
...
+    fp_add(fp_tmp[2], fp_tmp[4], fp_tmp[2]);             // U
```
Comment (SaveTheRbtz) [inline crypto/bls12381_hashtocurve.c:152]: nit: // U = b * V + (N^3 + a * N * D^2)

File: crypto/bls12381_hashtocurve.c [hunk 3]
```diff
@@ -142,95 +72,108 @@ const uint64_t iso_Ny_data[ELLP_Ny_LEN][Fp_DIGITS] = {
...
-    } else if (dv_cmp(bls_prec->fp_p_1div2, t, Fp_DIGITS) ==  RLC_LT) {
-        // g(X0(t)) was square and t is negative, so negate y
-        fp_neg_basic(fp_tmp[5], fp_tmp[5]);  // negate y because t is negative
+    int is_sqr = sqrt_ratio_3mod4(p->y, fp_tmp[2], fp_tmp[0]);
+    if (is_sqr) {
+        fp_copy(p->x, fp_tmp[3]);
+    } else {
+        fp_mul(p->x, fp_tmp[1], fp_tmp[3]);
+        fp_mul(fp_tmp[1], fp_tmp[1], t); 
+        fp_mul(p->y, p->y, fp_tmp[1]);
```
Comment (SaveTheRbtz) [inline crypto/bls12381_hashtocurve.c:161]: nit: math comments here

File: crypto/bls12381_hashtocurve.c [hunk 4]
```diff
@@ -142,95 +72,108 @@ const uint64_t iso_Ny_data[ELLP_Ny_LEN][Fp_DIGITS] = {
...
+// sqrt_ration optimized for p mod 4 = 3.
+// check if (U/V) is a square, return 1 if yes, 0 otherwise 
+// if 1 is returned, out contains sqrt(U/V)
+// out should not be the same as U, or V
+static int sqrt_ratio_3mod4(fp_t out, const fp_t u, const fp_t v) {
+    const int tmp_len = 3;
+    fp_t* fp_tmp = (fp_t*) malloc(tmp_len*sizeof(fp_t));
+    for (int i=0; i<tmp_len; i++) fp_new(fp_tmp[i]);
+
+    fp_sqr(fp_tmp[1], v);                               // V^2
```
[the remark this answers is not in the document]
Comment (SaveTheRbtz) [reply crypto/bls12381_hashtocurve.c:84]: in other words, can we do something like:
```
fp_t v2; fp_sqr(v2, v);  // V^2
fp_t uv; fp_mul(uv, u, v); // UV
fp_t uv3; fp_mul(uv3, uv, v2);  // UV^3
```

File: crypto/bls12381_hashtocurve.c [hunk 5]
```diff
@@ -142,95 +72,108 @@ const uint64_t iso_Ny_data[ELLP_Ny_LEN][Fp_DIGITS] = {
...
-    {0x09319f2e39834935, 0x039e952cbdb05c21, 0x55ba77a9a2f76493,
-     0xfd04e3dfc6086467, 0xfb95832e7d78742e, 0x0ef9c24eccaf5e0e,},
-};
+// sqrt_ration optimized for p mod 4 = 3.
+// check if (U/V) is a square, return 1 if yes, 0 otherwise 
+// if 1 is returned, out contains sqrt(U/V)
+// out should not be the same as U, or V
+static int sqrt_ratio_3mod4(fp_t out, const fp_t u, const fp_t v) {
+    const int tmp_len = 3;
+    fp_t* fp_tmp = (fp_t*) malloc(tmp_len*sizeof(fp_t));
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/bls12381_hashtocurve.c:81]: Makes sense 👌🏼

File: crypto/bls12381_hashtocurve.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (SaveTheRbtz) [reply crypto/bls12381_hashtocurve.c:84]: Thanks for the background -- it was quite an interesting read.  Maybe we can at least switch from an array to a list of variables like in the rfc draft?  This should give the compiler more flexibility (no need for a continuous allocations, (in some hardening modes) no bounds checking, etc).

**Comment by tarakby** [comment]
bors merge

File: crypto/bls12381_hashtocurve.c [hunk 6]
```diff
@@ -220,7 +212,6 @@ static inline void compute_map_zvals(const fp_t inv[], fp_t zv[],
 static inline void eval_iso11(ep_t r, const ep_t  p) {
     const int tmp_len = 32;
     fp_t* fp_tmp = (fp_t*) malloc(tmp_len*sizeof(fp_t));
```
Comment (SaveTheRbtz) [inline crypto/bls12381_hashtocurve.c:214]: nit: maybe replace this with on-stack `fp_t fp_tmp[32];`?

**Review by SaveTheRbtz — APPROVED** [review]

File: crypto/bls12381_hashtocurve.c (hunk 6 above)
Comment (SaveTheRbtz) [reply crypto/bls12381_hashtocurve.c:214]: Added as a part of onflow/flow-go#2598

**Pull request merged**